hazelcast/src/main/java/com/hazelcast/internal/util/collection/MPSCQueue.java

```diff
@@ -0,0 +1,319 @@
+     * to wakeup. There can only be a single consumerThread and this method should be called
+     * before the queue is safely published. It will not provide a happens before relation on
+     * its own.
+     *
+     * @param consumerThread the consumer thread.
+     * @throws NullPointerException when consumerThread null.
+     */
+    public void setConsumerThread(Thread consumerThread) {
+        this.consumerThread = checkNotNull(consumerThread, "consumerThread can't be null");
+    }
+
+    /**
+     * {@inheritDoc}.
+     *
+     * This call is threadsafe; but it will only remove the items that are on the put-stack.
+     */
+    @Override
+    public void clear() {
+        putStack.set(BLOCKED);
+    }
+
+    @Override
+    public boolean offer(E item) {
+        checkNotNull(item, "item can't be null");
+
+        AtomicReference<Node> putStack = this.putStack;
+        Node newHead = new Node();
+        newHead.item = item;
+
+        for (; ; ) {
+            Node oldHead = putStack.get();
+            if (oldHead == null || oldHead == BLOCKED) {
+                newHead.next = null;
+                newHead.size = 1;
+            } else {
+                newHead.next = oldHead;
+                newHead.size = oldHead.size + 1;
+            }
+
+            if (!putStack.compareAndSet(oldHead, newHead)) {
+                continue;
+            }
+
+            if (oldHead == BLOCKED) {
+                unpark(consumerThread);
+            }
+
+            return true;
+        }
+    }
+
+    @Override
+    public E take() throws InterruptedException {
+        E item = next();
+        if (item != null) {
+            return item;
+        }
+
+        takeAll();
+        assert takeStackIndex == 0;
+        assert takeStack[takeStackIndex] != null;
+
+        return next();
+    }
+
+    @Override
+    public E poll() {
+        E item = next();
+
+        if (item != null) {
+            return item;
+        }
+
+        if (!drainPutStack()) {
+            return null;
+        }
+
+        return next();
+    }
+
+    private E next() {
+        if (takeStackIndex == -1) {
+            return null;
+        }
+
+        if (takeStackIndex == takeStack.length) {
+            takeStackIndex = -1;
+            return null;
+        }
+
+        E item = (E) takeStack[takeStackIndex];
+        if (item == null) {
+            takeStackIndex = -1;
+            return null;
+        }
+
+        takeStack[takeStackIndex] = null;
+        takeStackIndex++;
+        takeStackSize.lazySet(takeStackSize.get() - 1);
+        return item;
+    }
+
+    private void takeAll() throws InterruptedException {
+        long iteration = 0;
+        AtomicReference<Node> putStack = this.putStack;
+        for (; ; ) {
+            if (consumerThread.isInterrupted()) {
+                putStack.compareAndSet(BLOCKED, null);
+                throw new InterruptedException("bla");
```

Comment:
oops.. remainder..  I was looking at some interrupt issues and trying to figure out what happened to mine. Later on I figured out it was caused by the unit test framework.